plugin-scalebar/src/main/java/com/mapbox/pluginscalebar/ScaleBarWidget.java

```diff
@@ -0,0 +1,214 @@
+    barPaint.setColor(primaryColor);
+    canvas.drawRect(margin - borderWidth, margin - borderWidth,
+      margin + unitBarWidth * pair.second + borderWidth, margin + barHeight + borderWidth, barPaint);
+
+    //Drawing the fill
+    barPaint.setStyle(Paint.Style.FILL);
+    int i = 0;
+    for (; i < pair.second; i++) {
+      barPaint.setColor(i % 2 == 0 ? primaryColor : secondaryColor);
+      String text = i == 0 ? String.valueOf(unitDistance * i) : unitDistance * i + unit;
+      canvas.drawText(text, margin + unitBarWidth * i, margin - barHeight, textPaint);
+      canvas.drawRect(margin + unitBarWidth * i, margin, margin + unitBarWidth * (1 + i),
+        margin + barHeight, barPaint);
+    }
+    canvas.drawText(unitDistance * i + unit, margin + unitBarWidth * i, margin - barHeight, textPaint);
+
+  }
+
+  /**
+   * Update the scale when mapView's scale has changed.
+   *
+   * @param metersPerPixel how many meters in each pixel.
+   */
+  void setDistancePerPixel(double metersPerPixel) {
+    this.distancePerPixel = isMetricSystem ? metersPerPixel : metersPerPixel * ScaleBarConstants.FEET_PER_METER;
+    if (!refreshHandler.hasMessages(MSG_WHAT)) {
+      refreshHandler.sendEmptyMessageDelayed(MSG_WHAT, REFRESH_DURATION);
+    }
+  }
+
+  /**
+   * Set the width of current mapView.
+   *
+   * @param mapViewWidth mapView's wdith.
+   */
+  void setMapViewWidth(int mapViewWidth) {
+    this.mapViewWidth = mapViewWidth;
+    margin = mapViewWidth / 20;
+    maxBarWidth = mapViewWidth / 2 - margin;
+  }
+
+  /**
+   * Set colors for the scale bar.
+   *
+   * @param textColor      The color for the texts above scale bar.
+   * @param primaryColor   The color for odd number index bars.
+   * @param secondaryColor The color for even number index bars.
+   */
+  void setColors(@ColorRes int textColor, @ColorRes int primaryColor, @ColorRes int secondaryColor) {
+    this.textColor = context.getResources().getColor(textColor);
+    this.primaryColor = context.getResources().getColor(primaryColor);
+    this.secondaryColor = context.getResources().getColor(secondaryColor);
+  }
+
+  @VisibleForTesting
+  public int getTextColor() {
+    return textColor;
+  }
+
+  @VisibleForTesting
+  public int getPrimaryColor() {
+    return primaryColor;
+  }
+
+  @VisibleForTesting
+  public int getSecondaryColor() {
+    return secondaryColor;
+  }
+
+  @VisibleForTesting
+  public int getMapViewWidth() {
+    return mapViewWidth;
+  }
+
+  /**
+   * Handler class to limit the refresh frequent.
+   */
+  private static class RefreshHandler extends Handler {
+    WeakReference<ScaleBarWidget> scaleBarWidgetWeakReference;
+
+    RefreshHandler(ScaleBarWidget scaleBarWidget) {
+      scaleBarWidgetWeakReference = new WeakReference<>(scaleBarWidget);
+    }
+
+    public void handleMessage(Message msg) {
+      if (msg.what == MSG_WHAT && scaleBarWidgetWeakReference.get() != null) {
```

Comment:
Could you move `scaleBarWidgetWeakReference.get()` up, out of the if-check, and store a reference in a local variable? Otherwise, you're still in risk of running into a race condition where the reference is valid during the if-check, but not during the access.

Reply:
Good point.